File: deploy/trinity-nexus/lang/libs/fsharp/TrinityVSA/TrinityVSA.fs
/// Trinity VSA - Vector Symbolic Architecture with balanced ternary arithmetic
module TrinityVSA

open System

type Trit = sbyte
type TritVector = Trit array

let zeros dim : TritVector = Array.zeroCreate dim

let random dim (seed: int) : TritVector =
    let rng = Random(seed)
    Array.init dim (fun _ -> sbyte (rng.Next(3) - 1))

let bind (a: TritVector) (b: TritVector) : TritVector =
    if a.Length <> b.Length then failwith "Dimension mismatch"
    Array.map2 (fun x y -> x * y) a b

let unbind a b = bind a b

let bundle (vectors: TritVector list) : TritVector =
    if List.isEmpty vectors then failwith "Empty vector list"
    let dim = vectors.[0].Length
    Array.init dim (fun i ->
        let sum = vectors |> List.sumBy (fun v -> int v.[i])
        if sum > 0 then 1y elif sum < 0 then -1y else 0y)

let permute (v: TritVector) shift : TritVector =
    let dim = v.Length
    let result = Array.zeroCreate dim
    for i in 0 .. dim - 1 do
        let newIdx = ((i + shift) % dim + dim) % dim
        result.[newIdx] <- v.[i]
    result

let dot (a: TritVector) (b: TritVector) : int64 =
    if a.Length <> b.Length then failwith "Dimension mismatch"
    Array.fold2 (fun acc x y -> acc + int64 x * int64 y) 0L a b

let similarity a b =
    let d = float (dot a b)
    let normA = sqrt (float (dot a a))
    let normB = sqrt (float (dot b b))
    if normA = 0.0 || normB = 0.0 then 0.0
    else d / (normA * normB)

let hammingDistance (a: TritVector) (b: TritVector) =
    if a.Length <> b.Length then failwith "Dimension mismatch"
    Array.fold2 (fun acc x y -> if x <> y then acc + 1 else acc) 0 a b

File: deploy/trinity-nexus/lang/libs/erlang/trinity_vsa/src/trinity_vsa.erl
%% Trinity VSA - Vector Symbolic Architecture with balanced ternary arithmetic
-module(trinity_vsa).
-export([zeros/1, random/2, bind/2, unbind/2, bundle/1, permute/2, dot/2, similarity/2, hamming_distance/2]).

zeros(Dim) ->
    lists:duplicate(Dim, 0).

random(Dim, Seed) ->
    rand:seed(exsss, {Seed, Seed, Seed}),
    [rand:uniform(3) - 2 || _ <- lists:seq(1, Dim)].

bind(A, B) when length(A) =:= length(B) ->
    lists:zipwith(fun(X, Y) -> X * Y end, A, B).

unbind(A, B) ->
    bind(A, B).

bundle(Vectors) when length(Vectors) > 0 ->
    Dim = length(hd(Vectors)),
    [begin
        Sum = lists:sum([lists:nth(I, V) || V <- Vectors]),
        if Sum > 0 -> 1; Sum < 0 -> -1; true -> 0 end
    end || I <- lists:seq(1, Dim)].

permute(V, Shift) ->
    Dim = length(V),
    [lists:nth(((I - 1 - Shift) rem Dim + Dim) rem Dim + 1, V) || I <- lists:seq(1, Dim)].

dot(A, B) when length(A) =:= length(B) ->
    lists:sum(lists:zipwith(fun(X, Y) -> X * Y end, A, B)).

similarity(A, B) ->
    D = dot(A, B),
    NormA = math:sqrt(dot(A, A)),
    NormB = math:sqrt(dot(B, B)),
    case {NormA, NormB} of
        {0.0, _} -> 0.0;
        {_, 0.0} -> 0.0;
        _ -> D / (NormA * NormB)
    end.

hamming_distance(A, B) when length(A) =:= length(B) ->
    length([1 || {X, Y} <- lists:zip(A, B), X =/= Y]).
